spock-core/src/main/java/org/spockframework/mock/runtime/IMockMaker.java

```diff
@@ -0,0 +1,317 @@
+   *
+   * <p>Note: the call to this method is guarded by a call to {@link #isMockable(IMockCreationSettings)}, so you do not need to check it twice.</p>
+   *
+   * @param settings the settings
+   * @return the new mock instance
+   * @throws CannotCreateMockException if this {@link IMockMaker} can't create such a mock.
+   */
+  default IStaticMock makeStaticMock(IMockCreationSettings settings) throws CannotCreateMockException {
+    throw new UnsupportedOperationException("Operation not implemented by " + getId() + ", but claimed capability MockMakerCapability.STATIC_METHOD. Settings: " + settings);
+  }
+
+  /**
+   * Checks if with the passed {@link IMockCreationSettings} the {@link #makeMock(IMockCreationSettings)} would be able to mock it.
+   *
+   * <p>Implementation note: YOu do not need to check for the capabilities ({@link #getCapabilities()}) this is done by the caller.
+   * </p>
+   *
+   * @param settings the settings
+   * @return result if it would be mockable or not, with this {@link IMockMaker}
+   */
+  IMockabilityResult isMockable(IMockCreationSettings settings);
+
+  /**
+   * Returns information about a mock object, or {@code null} it the object is no mock.
+   *
+   * <p>Implementation hint:
+   * An {@link IMockMaker} must provide the means to identify if a certain object is a mock, and need to return an attached {@link IMockObject}.
+   * If your mock class implement {@link org.spockframework.mock.ISpockMockObject}, you can simply return {@code null}.</p>
+   * </p>
+   *
+   * @param object a mock object
+   * @return information about the mock object or {@code null}
+   */
+  default IMockObject asMockOrNull(Object object) {
+    return null;
+  }
+
+  @Immutable
+  @FunctionalInterface
+  @Beta
+  interface IMockabilityResult {
+    IMockabilityResult MOCKABLE = new IMockabilityResult() {
+      @Override
+      public boolean mockable() {
+        return true;
+      }
+
+      @Override
+      public String nonMockableReason() {
+        return "";
+      }
+    };
+
+    /**
+     * @return {@code true} if the type is mockable.
+     */
+    default boolean mockable() {
+      return false;
+    }
+
+    /**
+     * @return the reason why the type is not mockable, otherwise empty {@link String}.
+     */
+    String nonMockableReason();
+  }
+
+  /**
+   * The {@link IMockCreationSettings} specifies what the {@link IMockMaker} shall mock
+   * in the methods {@link IMockMaker#makeMock(IMockCreationSettings)} or {@link IMockMaker#makeStaticMock(IMockCreationSettings)}.
+   *
+   * @since 2.4
+   */
+  @Beta
+  interface IMockCreationSettings {
+
+    /**
+     * @return The ID of the {@link IMockMaker} to use, otherwise {@code null} if any applicable {@link IMockMaker} shall be used.
+     */
+    @Nullable
+    String getMockMakerId();
+
+    Class<?> getMockType();
+
+    MockNature getMockNature();
+
+    List<Class<?>> getAdditionalInterface();
+
+    @Nullable
+    List<Object> getConstructorArgs();
```

Comment:
Because there is a difference between `null` and `empty`. Null marks that the constructor args feature was not used, and `empty` to use the zero-args constructor. This was already the case in the old code, I just pushed it into the interface and marked in correctly.
We would then need to change it to e.g.  `Optional<List<Object>>`.